dokka-integration-tests/gradle/src/main/kotlin/org/jetbrains/dokka/it/gradle/junit/TestedVersions.kt

```diff
@@ -0,0 +1,228 @@
+        composeGradlePlugin: String?,
+    ) : this(
+        kgp = SemVer(kgp),
+        gradle = SemVer(gradle),
+        dgp = SemVer(dgp),
+        agp = agp?.let { SemVer(it) },
+        composeGradlePlugin = composeGradlePlugin?.let { SemVer(it) },
+    )
+
+    /**
+     * Check if the combination of versions is valid.
+     *
+     * For example, AGP 7 must use Gradle 7. If [agp] is `8.0.0` and [gradle] is `7.0.0`
+     * then the combination is invalid.
+     */
+    private fun isValid(): Boolean {
+        if (agp != null) {
+            if (agp.major != gradle.major) {
+                return false
+            }
+
+            when (kgp) {
+                SemVer("1.9.0") -> {
+                    if (agp !in SemVer("7.0.0")..SemVer("8.2.0")) {
+                        return false
+                    }
+                }
+            }
+        }
+
+        return true
+    }
+
+    private val versionsMap: Map<String, SemVer> =
+        buildMap {
+            put("Dokka", dgp)
+            put("Kotlin", kgp)
+            put("Gradle", gradle)
+            if (agp != null) put("Android", agp)
+            if (composeGradlePlugin != null) put("Compose", composeGradlePlugin)
+        }
+
+    /**
+     * Join the major and minor versions of each version.
+     * Used by [allTestedVersions] to filter out duplicated patch versions.
+     */
+    private val majorAndMinorVersions: String =
+        versionsMap
+            .mapValues { (_, v) -> "${v.major}.${v.minor}" }
+            .entries
+            .map { (k, v) -> "$k: $v" }
+            .sorted().joinToString(" ")
+
+    /**
+     * Join the versions in a dash-separated-string, for use in file names.
+     */
+    val dashSeparated: String =
+        versionsMap.entries
+            .map { (key, value) -> "$key-$value" }
+            .sorted()
+            .joinToString("_")
+
+    /**
+     * Pretty display name, for use in JUnit test names.
+     */
+    val displayName: String =
+        versionsMap.entries
+            .map { (key, value) -> "$key: $value" }
+            .sorted()
+            .joinToString(", ")
+
+    /**
+     * Compares every version, in order.
+     */
+    override fun compareTo(other: TestedVersions): Int {
+        versionsMap.entries
+            .sortedBy { it.key }
+            .forEach { (key, thisVersion) ->
+                val otherVersion = other.versionsMap[key] ?: return@forEach
+                val cmp = thisVersion.compareTo(otherVersion)
+                if (cmp != 0) {
+                    return cmp
+                }
+            }
+        return 0
+    }
+
+    companion object {
+
+        /** All possible Kotlin Gradle Plugin versions that could be tested. */
+        private val allKgpVersions = listOf(
+            "1.9.0", // required for Compose 1.5.1
+            "1.9.23", // required for Compose 1.5.12
+            "1.9.24", // required for Compose 1.5.14
+            "1.9.25",
+            "2.0.20",
```

Comment:
if there is both 2.0.20 and 2.0.21, then proabably 2.0.0 and 2.0.10 should be here—otherwise, I think it should be fine to have just one version per major Kotlin version (except if we have some specific issues there)

Reply:
Agreed, I removed 2.0.20, it wasn't necessary.